Make `singletons-th` compile with `mtl-2.3.*`

Currently, `singletons-th` fails to build with the latest Hackage release of `mtl`:

```
$ cabal build singletons-th --constraint="mtl==2.3.*" --constraint="transformers==0.6.*" -w ghc-9.2
Build profile: -w ghc-9.2.2 -O1
In order, the following will be built (use -v for more details):
 - singletons-th-3.1 (lib) (first run)
Preprocessing library for singletons-th-3.1..
Building library for singletons-th-3.1..
[ 3 of 29] Compiling Data.Singletons.TH.Util ( src\Data\Singletons\TH\Util.hs, C:\\Users\winferno\Documents\Hacking\Haskell\singletons\dist-newstyle\build\x86_64-windows\ghc-9.2.2\singletons-th-3.1\build\Data\Singletons\TH\Util.o )

src\Data\Singletons\TH\Util.hs:345:48: error:
    Not in scope: type constructor or class ‘ExceptT’
    |
345 | newtype UnravelM a = UnravelM { runUnravelM :: ExceptT String (Reader Bool) a }
    |                                                ^^^^^^^
```

I've revised `th-desugar` on Hackage to have `mtl < 2.3` upper bounds to mitigate the damage. Given that the breakage involves `ExceptT`, we may want to first wait for a new Hackage release of `mtl` that includes a fix for haskell/mtl#117.

--- a/singletons-th/CHANGES.md
+++ b/singletons-th/CHANGES.md
@@ -6,6 +6,7 @@ Changelog for the `singletons-th` project
 * Fix a bug in which the `singDecideInstances` and `showSingInstances`, as well
   as `deriving Show` declarations, would not respect custom
   `promotedDataTypeOrConName` options.
+* Allow building with `mtl-2.3.*`.
 
 3.1.1 [2022.08.23]
 ------------------

--- a/singletons-th/singletons-th.cabal
+++ b/singletons-th/singletons-th.cabal
@@ -54,7 +54,7 @@ library
   hs-source-dirs:     src
   build-depends:      base             >= 4.17 && < 4.18,
                       containers       >= 0.5,
-                      mtl              >= 2.2.1,
+                      mtl              >= 2.2.1 && < 2.4,
                       ghc-boot-th,
                       singletons       == 3.0.*,
                       syb              >= 0.4,

--- a/singletons-th/src/Data/Singletons/TH/Single/Monad.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single/Monad.hs
@@ -26,8 +26,10 @@ import Data.Singletons.TH.Promote.Monad ( emitDecs, emitDecsM )
 import Data.Singletons.TH.Util
 import Language.Haskell.TH.Syntax hiding ( lift )
 import Language.Haskell.TH.Desugar
-import Control.Monad.Reader
-import Control.Monad.Writer
+import Control.Monad ( liftM2 )
+import Control.Monad.IO.Class ( MonadIO )
+import Control.Monad.Reader ( MonadReader(..), ReaderT(..), asks )
+import Control.Monad.Writer ( MonadWriter, WriterT(..) )
 import Control.Applicative
 
 -- environment during singling

--- a/singletons-th/src/Data/Singletons/TH/Util.hs
+++ b/singletons-th/src/Data/Singletons/TH/Util.hs
@@ -16,10 +16,12 @@ import Language.Haskell.TH ( pprint )
 import Language.Haskell.TH.Syntax hiding ( lift )
 import Language.Haskell.TH.Desugar
 import Data.Char
-import Control.Monad hiding ( mapM )
-import Control.Monad.Except hiding ( mapM )
-import Control.Monad.Reader hiding ( mapM )
-import Control.Monad.Writer hiding ( mapM )
+import Control.Monad ( liftM, unless, when )
+import Control.Monad.Except ( ExceptT, runExceptT, MonadError(..) )
+import Control.Monad.IO.Class ( MonadIO )
+import Control.Monad.Reader ( MonadReader(..), Reader, ReaderT(..) )
+import Control.Monad.Trans ( MonadTrans )
+import Control.Monad.Writer ( MonadWriter(..), WriterT(..), execWriterT )
 import qualified Data.Map as Map
 import Data.Map ( Map )
 import Data.Bifunctor (second)

